### DISCUSSION TITLE
Looking to contribute

### DISCUSSION THREAD

**Discussion by kevin-pease** [qutebrowser/qutebrowser#7263] [General]
Hey people,

I'm currently doing a uni course in Software Development, and one of the final assignments is to contribute something to an active repo of considerable size. Qutebrowser seems to fit the bill because of the clear structure, friendly community and the fact that the majority of the code is in Python. However, since this will be my first contribution ever, I'm having great difficulties in picking an existing issue, or coming up with an issue of my own. I've read all the documentation. Would Qutebrowser be a suitable project for me to work on?

**Comment by The-Compiler** [comment]
Welcome! :tada: 

I think it's always best to contribute to a project which you already are using, or could at least see yourself using. Before attempting to contribute to a project (other than, say, trivial typo fixes), you should be an user of it, in my opinion.

From your description, I'm not quite sure if that's the case - if it isn't, in your position, I'd prefer something you are using already. If that's not an option, you will need to take extra care to ensure that you understand a given issue from the user's perspective before you look at it from a developer's perspective. (There are of course exceptions for this, for changes which are not user-facing - but those often require a more intricate understanding of the code in my experience).

For qutebrowser in particular, note that it has a quite big [PR backlog](https://github.com/qutebrowser/qutebrowser/projects/4), as the amount of people contributing code is usually much bigger than the amount of people reviewing/testing/maintaining those contributions. Thus, there can be some delay until your change(s) are looked at. If your university course requires that your changes be reviewed or even integrated, then I'd ask you to look elsewhere (but then again, if that's the case, that university course is probably a doing a disservice to open source projects).

With all that out of the way: qutebrowser has a [list of issues marked "easy"](https://github.com/qutebrowser/qutebrowser/issues?q=is%3Aopen+is%3Aissue+label%3Aeasy) which might be good for a first-time contributor to start on. Mind you, it's incredibly difficult even with lots of experience to foresee how much effort implementing something will be, but that would probably be a good starting point.

I'd suggest picking one of those, ideally a somewhat recent one which doesn't already show an assignee or associated pull request:

![image](https://user-images.githubusercontent.com/625793/174294775-9d970cda-fa6b-44cc-a37c-80e13ee04d13.png)

Especially for older ones, but also in case you have any questions, I'd really appreciate if you left a comment in the issue first, so I can read through it and provide some thoughts first.

Hope that helps! Feel free to ask if you encounter any issues - there's also an [IRC channel](https://github.com/qutebrowser/qutebrowser/blob/master/doc/help/index.asciidoc#getting-help) to talk to people in real-time.

**Reply by kevin-pease** [reply]
Thank you so much for your elaborate answer! I've been using Qutebrowser off and on, but not too extensively. Also, my uni doesn't require any commits to be merged; my contributions will be examined and graded by my teacher(s), so all in all Qutebrowser fits the bill. I just find it surprisingly hard to pick something and dive right in, hence my question. I'll reconsider everything and probably discuss it with my teachers and peers. In the meantime, thanks again for your answer. You've given me some great tips and pointers! :)

**Reply by The-Compiler** [reply]
Yay, great to hear! Let me know if I can help somehow!
